packages/astro/src/core/app/index.ts

```diff
@@ -280,12 +284,21 @@ export class App {
 					status
 				);
 				const page = (await mod.page()) as any;
-				if (mod.onRequest) {
+				if (runMiddleware && mod.onRequest) {
 					this.#pipeline.setMiddlewareFunction(mod.onRequest as MiddlewareEndpointHandler);
 				}
+				if (!runMiddleware) {
+					// make sure middleware set by other requests is cleared out
+					this.#pipeline.unsetMiddlewareFunction();
+				}
 				const response = await this.#pipeline.renderRoute(newRenderContext, page);
 				return this.#mergeResponses(response, originalResponse);
-			} catch {}
+			} catch {
+				// Middleware may be the cause of the error, so we try rendering 404/500.astro without it.
+				if (runMiddleware && mod.onRequest) {
+					return this.#renderError(request, { status, response: originalResponse, runMiddleware: false });
+				}
```

Comment:
Uhm, I vaguely remember that in dev, we run the middleware for 404, by I might be wrong. If I am correct, wouldn't this cause misalignment in behaviour?

Reply:
Not sure, I'll have to check.

Reply:
It seems to me that here we don't run the middleware when we render a 404. Did I get it wrong?

Reply:
If there is a 404.astro, we would. Both during production SSR, and dev.

To summarize the difference introduced by this PR, when the 404.astro or 500.astro rendering throws an error...
- dev: the dev overlay with the error is shown immediately.
- prod: the error page is attempted again without middleware.